tools/changelog/main.go

```diff
@@ -0,0 +1,218 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"bytes"
+	"fmt"
+	"log"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+
+	"github.com/alecthomas/kingpin/v2"
+	"github.com/gravitational/trace"
+
+	"github.com/gravitational/shared-workflows/libs/gh"
+	"github.com/gravitational/shared-workflows/libs/git"
+)
+
+var (
+	baseBranch = kingpin.Flag(
+		"base-branch",
+		"The base release branch to generate the changelog for.  It will be of the form branch/v*",
+	).Envar("BASE_BRANCH").String()
+
+	baseTag = kingpin.Flag(
+		"base-tag",
+		"The tag/version to generate the changelog from. It will be of the form vXX.Y.Z, e.g. v15.1.1",
+	).Envar("BASE_TAG").String()
+)
+
+func main() {
+	kingpin.Parse()
+
+	if err := prereqCheck(); err != nil {
+		log.Fatal(err)
+	}
+
+	workDir, err := os.Getwd()
+	if err != nil {
+		log.Fatal(trace.Wrap(err, "failed to get working directory"))
+	}
+
+	topDir, err := git.RunCmd(workDir, "rev-parse", "--show-toplevel")
+	if err != nil {
+		log.Fatal(err)
+	}
+	entDir := filepath.Join(topDir, "e")
+
+	// Figure out the branch and last version released for that branch
+	branch, err := getBranch(*baseBranch, workDir)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	lastVersion, err := getLastVersion(*baseTag, workDir)
+	if err != nil {
+		log.Fatal(trace.Wrap(err, "failed to determine last version"))
+	}
+
+	// Determine timestamps of releases which is used to limit Github search
+	timeLastRelease, timeLastEntRelease, timeLastEntMod, err := getTimestamps(topDir, entDir, lastVersion)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	// Generate changelogs
+	ossCLGen := &changelogGenerator{
+		isEnt: false,
+		dir:   topDir,
+	}
+	entCLGen := &changelogGenerator{
+		isEnt: true,
+		dir:   entDir,
+	}
+	ossCL, err := ossCLGen.generateChangelog(branch, timeLastRelease, timeNow)
+	if err != nil {
+		log.Fatal(err)
+	}
+	entCL, err := entCLGen.generateChangelog(branch, timeLastEntRelease, timeLastEntMod)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	fmt.Println(ossCL)
+	if entCL != "" {
+		fmt.Println("Enterprise:")
+		fmt.Println(entCL)
+	}
+}
+
+func prereqCheck() error {
+	if err := git.IsAvailable(); err != nil {
+		return trace.Wrap(err)
+	}
+	if err := gh.IsAvailable(); err != nil {
+		return trace.Wrap(err)
+	}
+	return nil
+}
+
+// getBranch will return branch if parsed otherwise will attempt to find it
+// Branch should be in the format "branch/v*"
+func getBranch(branch, dir string) (string, error) {
```

Comment:
Can these things (get branch, get forked branch, etc.) be moved into a shared reusable "git" library package, similar to what we have for "github"?

Reply:
At that point it might just be easier to use https://github.com/go-git/go-git and remove the CLI tool dep entirely at the same time. I've used that library on a handful of projects before and it works well

Reply:
Yep, agreed.

Reply:
Yup I can do that should be a quick change.

Reply:
fwiw - Be sure to check the go-git [compatibility guide](https://github.com/go-git/go-git/blob/master/COMPATIBILITY.md). I had to go with the [git lib](https://github.com/gravitational/cloud/blob/master/tcctl/release/git/git.go) route for a different project due to a lack of support for certain commands.